### DISCUSSION TITLE
Built-in HTML parser and DOM

### DISCUSSION THREAD

**Discussion by vjpr** [oven-sh/bun#1522] [Ideas]
I wonder if we could offer built-in support for an HTML parser and the DOM.

# Use cases

- Web scraping
  - Quite common, and perf is important here because it's usually done at scale.
- Testing
  - A lot of testing is done in a Node.js environment using js-dom. Speeding up test suites is always cool.
  - With SSR on the rise, this is becoming more important, rather than traditionally relying on client-side E2E testing in a real browser or wrapper such as Cypress or using Puppeteer or Playwright.

# Why Bun should do this?

- Bun already has WebKit in its source tree. We could just re-use the existing implementations there. E.g. https://github.com/WebKit/webkit/blob/main/Source/WebCore/dom/DOMImplementation.cpp / https://github.com/WebKit/webkit/blob/main/Source/WebCore/html/parser/HTMLDocumentParser.cpp

- Demonstrating perf benefits for web scraping and DOM testing could push adoption.

- Seamless access to web platform from server. Currently you have to think: "am I in browser env or node env?" when doing web stuff.

# Existing HTML/DOM Libraries

A lot is built around htmlparser2: https://github.com/fb55/htmlparser2/#ecosystem

## https://cheerio.js.org/ - parse html and run jquery-like selectors

> Cheerio wraps around [parse5](https://github.com/inikulin/parse5) parser and can optionally use @FB55's forgiving [htmlparser2](https://github.com/fb55/htmlparser2/) (TypeScript). Cheerio can parse nearly any HTML or XML document.

## https://github.com/fb55/htmlparser2

> htmlparser2 is the fastest HTML parser

See: https://github.com/fb55/htmlparser2/#performance

> If you need strict HTML spec compliance, have a look at parse5.

## https://github.com/inikulin/parse5

> As of August 2021, parse5 is the only fully spec compliant HTML parser available for NodeJS
> From: https://github.com/AndreasMadsen/htmlparser-benchmark

## Benchmarks

Here are the benchmarks to beat: 

- https://github.com/AndreasMadsen/htmlparser-benchmark
  - Current results: https://github.com/fb55/htmlparser2/#performance

# Existing Testing Libraries

### Custom environments

Jest allows you to choose an env (node vs jsdom) or use a custom one 

https://jestjs.io/docs/configuration#testenvironment-string

### Sandboxing

> TestEnvironment is sandboxed. Each test suite will trigger setup/teardown in their own TestEnvironment.

> Jest creates a new jsdom environment for each test

Many testing libraries do this which could be quite processor intensive - hence more opportunities for perf wins.

# Renderer

If we bring in the parser and DOM now, in addition to JS engine, the only thing left now is the renderer.

After bringing in parsing/dom, the two parts left would be:

- Headless browser
  - already quite a crowded space (https://github.com/dhamaniasad/HeadlessBrowsers)
- Desktop browser
  - Bun replacement for Electron/Tauri. 
    - A layer for Bun.js to talk to a browser renderer.
    - Existing discussion: https://github.com/oven-sh/bun/discussions/790
  - Testing browser
    - Cypress is pretty cool - a desktop Chromium wrapper with a nice driver api.
      - It doesn't feel as integrated as it could be though.
      - I feel like the goal should be: "here is a chunk of TypeScript code, and here is where I want to run, and run it as fast as possible".

I guess the **advantage of doing a browser in Bun** would be to offer seamless interaction with the entire web platform. At present it's quite patchwork - there is a lot of existing native WebKit code that has been re-implmented (slower) in userland JS. And to access browser environments you need a lot of additonal libraries.

**Comment by vjpr** [comment]
tricked:

> bun has this in deps https://github.com/cloudflare/lol-html so i would think its maybe somewhere already not sure tho

---

jarred:

> LOLHTML doesn’t really have a full-fledged DOM
> It’s designed for streaming and rewriting
> rather than ASTs
> I think something in here is a good ide a
> But probably not for awhile
> CSS is higher priority

**Comment by mquandalle** [comment]
Deno has https://deno.land/manual@v1.36.4/advanced/jsx_dom/deno_dom

**Comment by ethanchristensen01** [comment]
I was using HTMLRewriter (LOLHTML) to work with an XML document and ran into problems when the document had Style tags. Children of Style were processed as raw text instead of as elements. I would like to work with LOLHTML to add support for xml documents.

**Comment by The-Code-Monkey** [comment]
@vjpr did you find a solution for this as I need to use parseFromString from DOMParser but cannot as it doesn't exist in bun.

**Comment by guest271314** [comment]
There is JSDOM https://github.com/guest271314/jsdom-extension.

**Comment by The-Code-Monkey** [comment]
How would I use JSDOM to parse a html string into html elements

**Reply by guest271314** [reply]
HTML elements where, in what environment?

**Comment by The-Code-Monkey** [comment]
In the browser, I have a rich text editor, that requires html code but it
stores the data as a string of that code, so I need to parse e.g. "<p>some
text<p>" to an actual html element. Which is what DOMParser.parseFromString
does, but DOMParser isn't part of bun yet, so I need some way to do this.

**Comment by guest271314** [comment]
@The-Code-Monkey Have you tried using the JSDOM code that I compiled for use in a browser extension `ServiceWorker`, linked above? Should work.

**Comment by The-Code-Monkey** [comment]
I need a browser based solution not a node one.

**Reply by guest271314** [reply]
You can run that code in the browser. That's what I wrote it for. It's JSDOM bundled to run in the browser. 

See https://github.com/guest271314/jsdom-extension/blob/main/background.js

```
// https://github.com/jsdom/jsdom/issues/3315
// https://github.com/GoogleChrome/developer.chrome.com/issues/2290
// https://github.com/jsdom/jsdom/issues/245#issuecomment-256615808
import {JSDOM} from './jsdom.js';         
let dom = new JSDOM(`<!DOCTYPE html><p>Hello world</p>`);
console.log(dom, dom.window.document.querySelector("p").textContent); // "Hello world"
```

**Reply by guest271314** [reply]
If you need to run the code in the browser you already have `DOMParser()`.

**Reply by The-Code-Monkey** [reply]
DOMParser doesn't work with bun as far as I could tell unless I just have my typescript settings wrong

**Reply by guest271314** [reply]
> DOMParser doesn't work with bun

I thought the target environment you are runningthe code in is the browser 

> I need a browser based solution not a node one.

**Reply by guest271314** [reply]
The bundled JSDOM script should work in the browser or in a non-browser JavaScript environment.

**Comment by vinyll** [comment]
Upgrading [Lego](https://github.com/Polight/lego) to V2 we are trying to minimize external dependencies.
Bun was favorited to achieve this, however it is missing a dom parser [like Deno has](https://doc.deno.land/deno/dom/~/DOMParser).
Is there any plan to embed this in the short/mid term?

**Comment by 21e8** [comment]
+1 would love this

**Comment by user451421541757324** [comment]
> CSS is higher priority

I was thinking whether, since the CSS feature was implemented with Bun 1.2, we could think about a DOM parser.

**Comment by MrDanielHarka** [comment]
+1! Would be nice!

**Comment by jamesgarrick** [comment]
+1

**Comment by mefistofelix** [comment]
For a modern native dom/xpath/xslt implementation instead of libxml2 which supports only xpath 1.0 see 
https://blog.startifact.com/posts/xee/

About an html5 parser the latest versions of PHP are using 
https://github.com/kostya/lexbor

Also adding a builtin json query language will be really cool, the most powerfull one (also used in the aws stack in some place) is:
https://github.com/jsonata-js/jsonata
Will be cool to try use some ai to port the JS to zig and jscore object access to have a builtin native implementation, because is a JS only lib

This additions will be really cool imho, Not only for web scraping but are a super powerfull toolbox for a wide range of common requirements nowdays, in one way or another.

**Comment by mefistofelix** [comment]
If using dom and HTML parser from WebKit i think you will inherit the dom event system, the mutation observer etc in a more native way than the one exposed in browsers to js... This could open to implementing in bun some interesting GUIs (os native or Not) that have the dom as underlying markup and source of truth/state/GUI/game scene graph...something like xaml and qml (Qt)

**Comment by peterlustig78** [comment]
as `bun build` is able to read html files for used script imports and other files to properly bundle them, bun should be able to parse it already using some internal feature. 
Not sure how it does it, and if it is useable in JS code.

**Comment by handylib** [comment]
+1

**Comment by bedros-p** [comment]
Why do we have S3 client magic and image processing baked into the runtime and not this? Feels like a major oversight to not have basic DOM support outside of HTMLRewriter
+1
